src/main/java/org/truffleruby/language/SafepointManager.java

```diff
@@ -230,15 +230,21 @@ public void pauseAllThreadsAndExecuteFromNonRubyThread(boolean deferred, Safepoi
     // Variants for a single thread
 
     @TruffleBoundary
-    public void pauseThreadAndExecute(Thread thread, Node currentNode, SafepointAction action) {
-        if (Thread.currentThread() == thread) {
+    public void pauseRubyThreadAndExecute(DynamicObject rubyThread, Node currentNode, SafepointAction action) {
+        final DynamicObject currentThread = context.getThreadManager().getCurrentThread();
+        final FiberManager fiberManager = Layouts.THREAD.getFiberManager(rubyThread);
+
+        if (currentThread == rubyThread) {
+            if (fiberManager.getRubyFiberFromCurrentJavaThread() != fiberManager.getCurrentFiber()) {
+                throw new AssertionError("Currently executing thread is not the current Fiber of the current Ruby Thread");
```

Comment:
Why explicitly throw an `AssertionError`? I'd rather it be an `assert` or `UnsupportedOperationException` like we use elsewhere.

Reply:
I've read this a few times and I can't really make sense of the error message.

Reply:
We only expect one active Fiber per Thread, and the initiating thread of execution should be a current Fiber. It seems useful to check this here as we rely on this invariant.
`UnsupportedOperationException` does not seem appropriate for a broken invariant that I want to always check.

Reply:
Truffle also does this explicit throwing of `AssertionError` and it leads to confusion when you haven't explicitly enabled assertions. I often wonder if `-ea` has been passed somewhere in our chain of scripts, before realizing it's being done with an explicit throw.

Reply:
Agreed on that, I had the same thoughts about it. I'll switch to a `UnsupportedOperationException`.

Reply:
Thanks for the explanation. I understand it better now. Maybe a slightly improved message would be: "The currently executing Java thread does not correspond to the currently active fiber for the current Ruby thread" or "The currently executing Java thread does not belong to the currently active fiber for the current Ruby thread". I think that clarifies a bit more. But do whatever you think is best.

Reply:
"The currently executing Java thread does not correspond to the currently active fiber for the current Ruby thread" sounds good, thanks!

Reply:
IllegalStateException or IllegalArgumentException might be better, unsupported suggests that it might be implemented in future.

Reply:
IllegalStateException sounds nice for "broken invariant", so I went with that.